tools/data_gen/src/main.rs

```diff
@@ -0,0 +1,197 @@
+    /// file path to read data from
+    #[argh(positional, from_str_fn(read_path))]
+    file: Option<PathBuf>,
+}
+
+fn read_path(path: &str) -> Result<PathBuf, String> {
+    Ok(Path::new(path).into())
+}
+
+fn main() -> CalyxResult<()> {
+    let fp: FilePath = argh::from_env();
+
+    let lib_path: &Path = &Path::new("../../../calyx");
+
+    let ws = frontend::Workspace::construct(&fp.file, lib_path)?;
+    let ctx: ir::Context = (ir::from_ast::ast_to_ir(ws))?;
+
+    let comp = ctx
+        .components
+        .into_iter()
+        .find(|comp| comp.name == ctx.entrypoint)
+        .expect("No top-level component found.");
+
+    let data_vec: Vec<CellData> = comp
+        .cells
+        .iter()
+        .filter_map(|cell| get_data(cell))
+        .collect();
+
+    let mut map = Map::new();
+
+    for CellData {
+        name,
+        width,
+        mut sizes,
+    } in data_vec
+    {
+        let json_comp: serde_json::Value = gen_comp(&mut sizes, width);
+        map.insert(name, json_comp);
+    }
+
+    let json_map: Value = map.into();
+    println!("{}", json_map.to_string());
+    Ok(())
+}
+
+//generates a json value associated with sizes_vec and width
+fn gen_comp(sizes_vec: &mut Vec<u64>, width: u64) -> serde_json::Value {
+    let initial: SizeVec = SizeVec::Single(vec![
+        0;
+        sizes_vec
+            .pop()
+            .unwrap_or_else(|| panic!("unable to read from size_vec"))
+            .try_into()
+            .unwrap_or_else(
+                |_| panic!("unable to turn u64 into usize")
+            )
+    ]);
+    sizes_vec.reverse();
+    let data_vec: SizeVec =
+        sizes_vec.iter().fold(initial, |acc, x| accumulate(*x, acc));
+    let data = serde_json::to_value(data_vec)
+        .unwrap_or_else(|_| panic!("could not create data array"));
+    json!({
+        "data": data.get("data").unwrap_or_else(|| panic!("could place data array in json value")),
+        "format": {
+            "numeric_type": "bitnum",
+            "is_signed": false,
+            "width": width,
+        }
+    })
+}
+
+//creates a vector consisting of i v's.
+fn add_dimension<T: std::clone::Clone>(i: u64, v: T) -> Vec<T> {
+    vec![
+        v;
+        i.try_into()
+            .unwrap_or_else(|_| panic!("could not add dimension to data array"))
+    ]
+}
+
+//function to help build vectors of multiple dimensions using fold()
+fn accumulate(i: u64, v: SizeVec) -> SizeVec {
+    match v {
+        SizeVec::Single(v) => SizeVec::Double(add_dimension(i, v)),
+        SizeVec::Double(v) => SizeVec::Triple(add_dimension(i, v)),
+        SizeVec::Triple(v) => SizeVec::Quad(add_dimension(i, v)),
+        SizeVec::Quad(_) => panic!("trying to add dimension to 4d"),
+    }
+}
+
+//Takes in cell, returns the reference to the cell
+fn get_ref(cell: &Rc<RefCell<ir::Cell>>) -> Ref<ir::Cell> {
+    let ref_cell: &RefCell<ir::Cell> = cell.borrow();
+    let final_cell: Ref<ir::Cell> = ref_cell.borrow();
+    final_cell
+}
```

Comment:
This function is unneeded. Once you remove the import for the borrow trait (see line 8) this will turn into `cell.borrow()`. The confusion comes from the fact that the call on 165 is the `borrow` from the `Borrow` trait, while the call on line 166 is the `borrow()` method of `RefCell`. Normally, because of the deref you can just call this directly from the rc, but because you have the borrow trait imported, it instead uses the borrow trait method on the RC itself, rather than using the RC's deref to the underlying RefCell and calling the RefCell's method.

Confusing, I know! Anyway, just the remove the line indicated (and maybe tell VSCode to not do auto-imports if you use it) and it should all wrap up without difficulty